core/rate_limiter.js

```diff
@@ -0,0 +1,72 @@
+var logging = require('minilog')('radar:rate_limiter');
+
+var RateLimiter = function(op, limit) {
+ this._op = op;
+ this._limit = limit;
+ this._resources = {};
+};
+
+RateLimiter.prototype.add = function(id, name, operation) {
+  if (this._op !== operation) {
+    return false;
+  }
+
+  if (!this._isNewResource(id, name)) {
+    return false;
+  }
+
+  if (this.isNearLimit(id)) {
+    logging.warn('rate limiting client: ' + id + ' name: ' + name);
+    return false;
+  }
+
+  this._add(id, name);
+  return true;
+};
+
+RateLimiter.prototype.isNearLimit = function(id) {
+  return this.count(id) >= this._limit;
+};
+
+RateLimiter.prototype.count = function(id) {
+  var resources = this._getResources(id);
+  return (resources ? resources.length : 0);
+};
+
+RateLimiter.prototype.reset = function(id) {
+  var resources = this._getResources(id);
+  delete resources;
+};
```

Comment:
Name this *remove* instead of *reset*, since we have *add* above?  Again, *resources_per_id* here ?